lxc-to-lxd/main_netcat.go

```diff
@@ -0,0 +1,72 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"sync"
+
+	"github.com/spf13/cobra"
+
+	"github.com/lxc/lxd/shared/eagain"
+)
+
+type cmdNetcat struct {
+	global *cmdGlobal
+}
+
+func (c *cmdNetcat) Command() *cobra.Command {
+	cmd := &cobra.Command{}
+
+	cmd.Use = "netcat <address>"
+	cmd.Short = "Sends stdin data to a unix socket"
+	cmd.RunE = c.Run
+	cmd.Hidden = true
+
+	return cmd
+}
+
+func (c *cmdNetcat) Run(cmd *cobra.Command, args []string) error {
+	// Help and usage
+	if len(args) == 0 {
+		cmd.Help()
+		return nil
+	}
+
+	// Handle mandatory arguments
+	if len(args) != 1 {
+		cmd.Help()
+		return fmt.Errorf("Missing required argument")
+	}
+
+	// Connect to the provided address
+	uAddr, err := net.ResolveUnixAddr("unix", args[0])
+	if err != nil {
+		return err
+	}
+
+	conn, err := net.DialUnix("unix", nil, uAddr)
+	if err != nil {
+		return err
+	}
+
+	// We'll wait until we're done reading from the socket
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	go func() {
+		io.Copy(eagain.Writer{Writer: os.Stdout}, eagain.Reader{Reader: conn})
+		conn.Close()
+		wg.Done()
+	}()
+
+	go func() {
+		io.Copy(eagain.Writer{Writer: conn}, eagain.Reader{Reader: os.Stdin})
```

Comment:
Are you sure this can't hang?

Reply:
Why should it?

Reply:
We had issues with the proxy device type and I had to fix a lot of stuff in there. This should verify that both go routines are actually terminating. The best way to do this would be to have a channel and then have both go routines write to that channel when they are finished like you do below.

Reply:
I'll take a look at updating the p2c code, then we can update lxc-to-lxd to match.

Reply:
It's less likely to be an issue than in the proxy case because we're actually dealing with a unidirectional stream (even though we forward both ways). Rsync just spews stuff through and doesn't really get any data back at all.

Reply:
I would prefer this to be changed to something like:
```go
diff --git a/lxc-to-lxd/main_netcat.go b/lxc-to-lxd/main_netcat.go
index db21bed71..34b4d0e22 100644
--- a/lxc-to-lxd/main_netcat.go
+++ b/lxc-to-lxd/main_netcat.go
@@ -5,7 +5,6 @@ import (
 	"io"
 	"net"
 	"os"
-	"sync"
 
 	"github.com/spf13/cobra"
 
@@ -52,21 +51,32 @@ func (c *cmdNetcat) Run(cmd *cobra.Command, args []string) error {
 	}
 
 	// We'll wait until we're done reading from the socket
-	wg := sync.WaitGroup{}
-	wg.Add(1)
+	chSend := make(chan error)
+	chRecv := make(chan error)
 
 	go func() {
-		io.Copy(eagain.Writer{Writer: os.Stdout}, eagain.Reader{Reader: conn})
-		conn.Close()
-		wg.Done()
+		_, err := io.Copy(eagain.Writer{Writer: os.Stdout}, eagain.Reader{Reader: conn})
+		chSend <- err
 	}()
 
 	go func() {
-		io.Copy(eagain.Writer{Writer: conn}, eagain.Reader{Reader: os.Stdin})
+		_, err := io.Copy(eagain.Writer{Writer: conn}, eagain.Reader{Reader: os.Stdin})
+		chRecv <- err
 	}()
 
-	// Wait
-	wg.Wait()
+	select {
+	case errSnd := <-chSend:
+		if errSnd != nil {
+			fmt.Printf("Error while sending data: %v\n", errSnd)
+		}
+
+	case errRcv := <-chRecv:
+		if errRcv != nil {
+			fmt.Printf("Error while reading data: %v\n", errRcv)
+		}
+	}
+
+	conn.Close()
 
 	return nil
 }
```